// @pipecat-ai/client-react/dist/index.js
var $5Zyvw$jotai = require("jotai");
var $5Zyvw$reactjsxruntime = require("react/jsx-runtime");
var $5Zyvw$jotaiutils = require("jotai/utils");
var $5Zyvw$react = require("react");
var $5Zyvw$pipecataiclientjs = require("@pipecat-ai/client-js");
var $5Zyvw$jotaireact = require("jotai/react");


function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

$parcel$export(module.exports, "deduplicateFunctionCalls", () => $cbb62fb16c5d45ff$export$dfa183f99573a822);
$parcel$export(module.exports, "filterBotOutputText", () => $62d3a839f4db69ec$export$ea20379ac23e22f2);
$parcel$export(module.exports, "filterEmptyMessages", () => $cbb62fb16c5d45ff$export$3d3951a55577516);
$parcel$export(module.exports, "isMessageEmpty", () => $cbb62fb16c5d45ff$export$f4610a88b89838b);
$parcel$export(module.exports, "mergeMessages", () => $cbb62fb16c5d45ff$export$23a96e868837e158);
$parcel$export(module.exports, "PipecatClientAudio", () => $a2a1c5f475114d56$export$b52250cb73ff4de1);
$parcel$export(module.exports, "PipecatClientCamToggle", () => $666ceba11158b494$export$dc9a029eeca8213f);
$parcel$export(module.exports, "PipecatClientMicToggle", () => $1a9931980e271aa3$export$bc8133b69ff660a2);
$parcel$export(module.exports, "PipecatClientProvider", () => $8df0e777e4d7dd49$export$bb43666ced7a20d0);
$parcel$export(module.exports, "PipecatClientScreenShareToggle", () => $0a9d1dc5be2e1cc2$export$93764714dfab3a46);
$parcel$export(module.exports, "PipecatClientVideo", () => $0f97689637ada1d8$export$85974db6d0cc43b3);
$parcel$export(module.exports, "sortByCreatedAt", () => $cbb62fb16c5d45ff$export$cbefc5865f8cbd89);
$parcel$export(module.exports, "useConversationContext", () => $984246ab4e966dee$export$8eec679bf8249822);
$parcel$export(module.exports, "usePipecatClient", () => $172f489fc5d91d99$export$777fa8498be78705);
$parcel$export(module.exports, "usePipecatClientCamControl", () => $d9b24817de62910a$export$3ea2601427f0430f);
$parcel$export(module.exports, "usePipecatClientMediaDevices", () => $9bd3e7d3a9d7acd1$export$642bc4d2d2a376f1);
$parcel$export(module.exports, "usePipecatClientMediaTrack", () => $630203d8dad1dd45$export$9813dcd2d0c26814);
$parcel$export(module.exports, "usePipecatClientMicControl", () => $4af0eba414c586fd$export$388e706586309ef0);
$parcel$export(module.exports, "usePipecatClientScreenShareControl", () => $5fc67d7ca05dec34$export$be63b19bd7f7d4f5);
$parcel$export(module.exports, "usePipecatClientTransportState", () => $810478f6ae107062$export$30aee278309a867b);
$parcel$export(module.exports, "usePipecatConversation", () => $fe6dfdd0b8560c78$export$acb832b064cb6c09);
$parcel$export(module.exports, "useRTVIClientEvent", () => $8a6b68ebf0332682$export$33a6ac53b8f02625);
$parcel$export(module.exports, "VoiceVisualizer", () => $a1dfa75b13e6bb9b$export$59bf27bd43679db6);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ /**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ /**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ const $ad0c8ea1ee293a79$export$7d7fdcf7579a70b2 = (text)=>{
    return text.toLowerCase().replace(/[^\p{L}\p{N}\s]/gu, "");
};
const $ad0c8ea1ee293a79$var$skipWhitespace = (text, start)=>{
    let i = start;
    while(i < text.length && /\s/.test(text[i]))i++;
    return i;
};
/**
 * Finds where `spoken` appears in `unspoken`, starting from `startPosition`.
 * - Best-effort sequential word matching (normalized, punctuation-stripped)
 * - Returns the original start position on mismatch (no advancement)
 */ const $ad0c8ea1ee293a79$var$findSpokenPositionInUnspoken = (spoken, unspoken, startPosition)=>{
    if (!spoken || !unspoken || startPosition >= unspoken.length) return startPosition;
    // If spoken text includes a leading separator space, skip leading whitespace in unspoken.
    let actualStart = startPosition;
    let spokenForMatching = spoken;
    if (spoken.startsWith(" ") && startPosition < unspoken.length) {
        actualStart = $ad0c8ea1ee293a79$var$skipWhitespace(unspoken, startPosition);
        spokenForMatching = spoken.trimStart();
    } else if (startPosition === 0 && startPosition < unspoken.length) // If we're at the start, also skip leading whitespace (e.g. newlines)
    actualStart = $ad0c8ea1ee293a79$var$skipWhitespace(unspoken, 0);
    const remainder = unspoken.slice(actualStart);
    // Sentence-level: if spoken exactly matches the remainder (normalized), consume the whole part
    // so we never leave a word unspoken due to word-matching edge cases.
    if ($ad0c8ea1ee293a79$export$7d7fdcf7579a70b2(spokenForMatching).trim() === $ad0c8ea1ee293a79$export$7d7fdcf7579a70b2(remainder).trim()) return unspoken.length;
    const spokenWords = $ad0c8ea1ee293a79$export$7d7fdcf7579a70b2(spokenForMatching).split(/\s+/).filter(Boolean);
    if (spokenWords.length === 0) return actualStart;
    const unspokenWords = $ad0c8ea1ee293a79$export$7d7fdcf7579a70b2(unspoken.slice(actualStart)).split(/\s+/).filter(Boolean);
    // Sequential match, allowing prefix match for contractions (e.g. "I" vs "I'm")
    // and limited skipping of mismatched unspoken words (e.g. punctuation artifacts).
    let matchedWords = 0;
    let consecutiveSkips = 0;
    let lastMatchedUnspokenIdx = -1;
    const MAX_CONSECUTIVE_SKIPS = 2;
    for(let i = 0; i < unspokenWords.length && matchedWords < spokenWords.length; i++){
        const target = spokenWords[matchedWords];
        const candidate = unspokenWords[i];
        if (candidate === target || candidate.startsWith(target)) {
            matchedWords++;
            consecutiveSkips = 0;
            lastMatchedUnspokenIdx = i;
            continue;
        }
        consecutiveSkips++;
        if (consecutiveSkips > MAX_CONSECUTIVE_SKIPS) return actualStart;
    // Skip this unspoken word and try matching the next one
    }
    if (matchedWords < spokenWords.length) return actualStart;
    // Convert the matched unspoken-word index back into a character position.
    // Walk whitespace-separated "chunks" that contain at least one alphanumeric
    // char (so pure-punctuation chunks like "—" are skipped — matching the
    // normalize+split tokenization used above). Stop at the chunk whose index
    // equals lastMatchedUnspokenIdx, then advance past any trailing whitespace
    // so the cursor sits at the start of the next chunk.
    const targetChunkCount = lastMatchedUnspokenIdx + 1;
    const isAlnum = (char)=>/[\p{L}\p{N}]/u.test(char);
    const isWs = (char)=>/\s/.test(char);
    let chunkCount = 0;
    let inChunk = false;
    let chunkHasAlnum = false;
    let i = actualStart;
    while(i < unspoken.length){
        const c = unspoken[i];
        if (isWs(c)) {
            if (inChunk && chunkHasAlnum) {
                chunkCount++;
                if (chunkCount === targetChunkCount) {
                    while(i < unspoken.length && isWs(unspoken[i]))i++;
                    return i;
                }
            }
            inChunk = false;
            chunkHasAlnum = false;
        } else {
            inChunk = true;
            if (isAlnum(c)) chunkHasAlnum = true;
        }
        i++;
    }
    return unspoken.length;
};
function $ad0c8ea1ee293a79$export$802e9b947136afbf(cursor, parts) {
    if (parts.length === 0) return false;
    for(let i = 0; i < parts.length; i++){
        if (typeof parts[i]?.text !== "string") continue;
        if (!cursor.partFinalFlags[i]) return true;
    }
    return false;
}
function $ad0c8ea1ee293a79$export$9b39fd8438081d6d(cursor, parts, spokenText) {
    if (parts.length === 0) return false;
    // Find the next part that should be spoken (skip parts already marked final/skipped)
    let partToMatch = cursor.currentPartIndex;
    while(partToMatch < parts.length && cursor.partFinalFlags[partToMatch])partToMatch++;
    if (partToMatch >= parts.length) return false;
    if (partToMatch > cursor.currentPartIndex) {
        cursor.currentPartIndex = partToMatch;
        cursor.currentCharIndex = 0;
    }
    const currentPart = parts[cursor.currentPartIndex];
    if (typeof currentPart.text !== "string") return false;
    const partText = currentPart.text;
    const startChar = cursor.currentCharIndex;
    // If the spoken text is pure punctuation (e.g. an em dash "—"), normalization
    // strips it to empty and word-matching would fail. Treat it as successfully
    // consumed so the cursor stays put and subsequent words continue matching.
    if ($ad0c8ea1ee293a79$export$7d7fdcf7579a70b2(spokenText).trim().length === 0) return true;
    const newPosition = $ad0c8ea1ee293a79$var$findSpokenPositionInUnspoken(spokenText, partText, startChar);
    const whitespaceEnd = $ad0c8ea1ee293a79$var$skipWhitespace(partText, startChar);
    if (newPosition > whitespaceEnd) {
        cursor.currentCharIndex = newPosition;
        if (newPosition >= partText.length) {
            cursor.partFinalFlags[cursor.currentPartIndex] = true;
            if (cursor.currentPartIndex < parts.length - 1) {
                cursor.currentPartIndex++;
                cursor.currentCharIndex = 0;
            }
        }
        return true;
    }
    // Intra-part scan-ahead recovery: if matching failed at the current position,
    // scan forward word-by-word within the same part. This prevents the cursor
    // from getting permanently stuck mid-part when a single word mismatch occurs
    // (e.g. TTS variation, punctuation boundary like `apexes."Sometimes`).
    if (startChar > 0) {
        const MAX_SCAN_WORDS = 8;
        let scanPos = startChar;
        for(let scan = 0; scan < MAX_SCAN_WORDS; scan++){
            // Advance past current word
            while(scanPos < partText.length && !/\s/.test(partText[scanPos]))scanPos++;
            // Advance past whitespace to next word
            while(scanPos < partText.length && /\s/.test(partText[scanPos]))scanPos++;
            if (scanPos >= partText.length) break;
            const retryPos = $ad0c8ea1ee293a79$var$findSpokenPositionInUnspoken(spokenText, partText, scanPos);
            const scanWsEnd = $ad0c8ea1ee293a79$var$skipWhitespace(partText, scanPos);
            if (retryPos > scanWsEnd) {
                cursor.currentCharIndex = retryPos;
                if (retryPos >= partText.length) {
                    cursor.partFinalFlags[cursor.currentPartIndex] = true;
                    if (cursor.currentPartIndex < parts.length - 1) {
                        cursor.currentPartIndex++;
                        cursor.currentCharIndex = 0;
                    }
                }
                return true;
            }
        }
    }
    // Mismatch recovery: try to find the spoken text in a later part.
    for(let nextPartIdx = cursor.currentPartIndex + 1; nextPartIdx < parts.length; nextPartIdx++){
        const nextPart = parts[nextPartIdx];
        if (typeof nextPart.text !== "string") continue;
        const match = $ad0c8ea1ee293a79$var$findSpokenPositionInUnspoken(spokenText, nextPart.text, 0);
        const nextWhitespaceEnd = $ad0c8ea1ee293a79$var$skipWhitespace(nextPart.text, 0);
        if (match > nextWhitespaceEnd) {
            // Mark skipped parts as final and jump to the matched part
            for(let i = cursor.currentPartIndex; i < nextPartIdx; i++)cursor.partFinalFlags[i] = true;
            cursor.currentPartIndex = nextPartIdx;
            cursor.currentCharIndex = match;
            return true;
        }
    }
    // If we're stuck at the start, mark the current part as skipped to avoid deadlock.
    if (startChar === 0 && cursor.currentPartIndex < parts.length - 1) {
        cursor.partFinalFlags[cursor.currentPartIndex] = true;
        cursor.currentPartIndex++;
        cursor.currentCharIndex = 0;
        return true;
    }
    return false;
}


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
const $28e6b30207a0392e$export$92a4076839a978d0 = (0, $5Zyvw$jotai.atom)([]);
const $28e6b30207a0392e$export$bd01e11bc95333a1 = (0, $5Zyvw$jotai.atom)(new Map());
const $28e6b30207a0392e$export$b0cd27059c88d0ab = (0, $5Zyvw$jotai.atom)(new Map());
const $28e6b30207a0392e$export$57b1a1df06a7e728 = (0, $5Zyvw$jotai.atom)(null);
const $28e6b30207a0392e$export$3365e5c5967ff793 = (0, $5Zyvw$jotai.atom)(new Map());


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ // ES2020-compatible polyfills for findLast / findLastIndex
function $f4fa178c5adc67bb$export$8855a8be7bd3e9f8(arr, predicate) {
    for(let i = arr.length - 1; i >= 0; i--){
        if (predicate(arr[i], i, arr)) return i;
    }
    return -1;
}
function $f4fa178c5adc67bb$export$296de88ccac4bedb(arr, predicate) {
    const idx = $f4fa178c5adc67bb$export$8855a8be7bd3e9f8(arr, predicate);
    return idx === -1 ? undefined : arr[idx];
}


/** Max time gap (ms) between consecutive same-role messages for merging. */ const $cbb62fb16c5d45ff$var$MERGE_WINDOW_MS = 30000;
/**
 * Unicode characters used by `FilterIncompleteTurns` on the server to mark
 * turn completion status. The LLM emits one of these as the very first
 * character of every response:
 *
 *   ✓ (U+2713) — complete turn
 *   ○ (U+25CB) — incomplete short
 *   ◐ (U+25D0) — incomplete long
 *
 * They must be stripped before the text reaches conversation state.
 */ const $cbb62fb16c5d45ff$var$TURN_COMPLETION_MARKERS = new Set([
    "\u2713",
    "\u25CB",
    "\u25D0"
]);
function $cbb62fb16c5d45ff$export$146b18879fa8cfd7(text) {
    if (text.length === 0) return text;
    if ($cbb62fb16c5d45ff$var$TURN_COMPLETION_MARKERS.has(text[0])) // Remove marker and optional single trailing space
    return text[1] === " " ? text.slice(2) : text.slice(1);
    return text;
}
const $cbb62fb16c5d45ff$export$cbefc5865f8cbd89 = (a, b)=>{
    return a.createdAt < b.createdAt ? -1 : a.createdAt > b.createdAt ? 1 : 0;
};
const $cbb62fb16c5d45ff$export$f4610a88b89838b = (message)=>{
    if (message.role === "function_call") return false;
    const parts = message.parts || [];
    if (parts.length === 0) return true;
    return parts.every((p)=>{
        if (typeof p.text === "string") return p.text.trim().length === 0;
        // Check BotOutputText objects
        if (typeof p.text === "object" && p.text !== null && "spoken" in p.text && "unspoken" in p.text) {
            const botText = p.text;
            return botText.spoken.trim().length === 0 && botText.unspoken.trim().length === 0;
        }
        // For ReactNode, consider it non-empty
        return false;
    });
};
const $cbb62fb16c5d45ff$export$3d3951a55577516 = (messages)=>{
    return messages.filter((message, index, array)=>{
        if (!$cbb62fb16c5d45ff$export$f4610a88b89838b(message)) return true;
        // For empty messages, keep only if no following non-empty message with same role
        const nextMessageWithSameRole = array.slice(index + 1).find((m)=>m.role === message.role && !$cbb62fb16c5d45ff$export$f4610a88b89838b(m));
        return !nextMessageWithSameRole;
    });
};
const $cbb62fb16c5d45ff$export$23a96e868837e158 = (messages)=>{
    const mergedMessages = [];
    for(let i = 0; i < messages.length; i++){
        const currentMessage = messages[i];
        const lastMerged = mergedMessages[mergedMessages.length - 1];
        const timeDiff = lastMerged ? Math.abs(new Date(currentMessage.createdAt).getTime() - new Date(lastMerged.createdAt).getTime()) : Infinity;
        const shouldMerge = lastMerged && lastMerged.role === currentMessage.role && currentMessage.role !== "system" && currentMessage.role !== "function_call" && timeDiff < $cbb62fb16c5d45ff$var$MERGE_WINDOW_MS;
        if (shouldMerge) mergedMessages[mergedMessages.length - 1] = {
            ...lastMerged,
            parts: [
                ...lastMerged.parts || [],
                ...currentMessage.parts || []
            ],
            updatedAt: currentMessage.updatedAt || currentMessage.createdAt,
            final: currentMessage.final !== false
        };
        else mergedMessages.push(currentMessage);
    }
    return mergedMessages;
};
const $cbb62fb16c5d45ff$var$statusPriority = {
    started: 0,
    in_progress: 1,
    completed: 2
};
const $cbb62fb16c5d45ff$export$dfa183f99573a822 = (messages)=>{
    const bestByToolCallId = new Map();
    const toRemove = new Set();
    for(let i = 0; i < messages.length; i++){
        const msg = messages[i];
        const tcid = msg.functionCall?.tool_call_id;
        if (msg.role !== "function_call" || !tcid) continue;
        const existingIdx = bestByToolCallId.get(tcid);
        if (existingIdx !== undefined) {
            const existingPriority = $cbb62fb16c5d45ff$var$statusPriority[messages[existingIdx].functionCall.status] ?? 0;
            const currentPriority = $cbb62fb16c5d45ff$var$statusPriority[msg.functionCall.status] ?? 0;
            if (currentPriority >= existingPriority) {
                toRemove.add(existingIdx);
                bestByToolCallId.set(tcid, i);
            } else toRemove.add(i);
        } else bestByToolCallId.set(tcid, i);
    }
    if (toRemove.size === 0) return messages;
    return messages.filter((_, i)=>!toRemove.has(i));
};
const $cbb62fb16c5d45ff$var$normalizeMessagesForUI = (messages)=>{
    return $cbb62fb16c5d45ff$export$23a96e868837e158($cbb62fb16c5d45ff$export$dfa183f99573a822($cbb62fb16c5d45ff$export$3d3951a55577516([
        ...messages
    ].sort($cbb62fb16c5d45ff$export$cbefc5865f8cbd89))));
};
// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------
const $cbb62fb16c5d45ff$var$callCallbacks = (callbacksMap, type, message)=>{
    callbacksMap.forEach((callbacks)=>{
        try {
            callbacks[type]?.(message);
        } catch (error) {
            console.error(`Error in ${type} callback:`, error);
        }
    });
};
function $cbb62fb16c5d45ff$export$f2820c5e040afe17(get, set, id, callbacks) {
    const map = new Map(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)));
    map.set(id, callbacks);
    set((0, $28e6b30207a0392e$export$b0cd27059c88d0ab), map);
}
function $cbb62fb16c5d45ff$export$46ae1ba121fdc956(get, set, id) {
    const map = new Map(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)));
    map.delete(id);
    set((0, $28e6b30207a0392e$export$b0cd27059c88d0ab), map);
}
function $cbb62fb16c5d45ff$export$f2434643f2abff11(_get, set) {
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), []);
    set((0, $28e6b30207a0392e$export$bd01e11bc95333a1), new Map());
    set((0, $28e6b30207a0392e$export$3365e5c5967ff793), new Map());
}
function $cbb62fb16c5d45ff$export$16fdb433d434f08(get, set, messageData) {
    const now = new Date();
    const message = {
        ...messageData,
        createdAt: now.toISOString(),
        updatedAt: now.toISOString()
    };
    const current = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
    const updatedMessages = [
        ...current,
        message
    ];
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(updatedMessages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageCreated", message);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$a05f96a2aa40873e(get, set, role, updates) {
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const lastMessageIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === role);
    if (lastMessageIndex === -1) return;
    const existing = messages[lastMessageIndex];
    const updatedMessage = {
        ...existing,
        ...updates,
        updatedAt: new Date().toISOString()
    };
    messages[lastMessageIndex] = updatedMessage;
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageUpdated", updatedMessage);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$78bd074993ab081f(get, set, role) {
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const lastMessageIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === role);
    if (lastMessageIndex === -1) return;
    const lastMessage = messages[lastMessageIndex];
    // Check if message is empty
    if ($cbb62fb16c5d45ff$export$f4610a88b89838b(lastMessage)) // Remove empty message only if it has no text in streams either
    messages.splice(lastMessageIndex, 1);
    else {
        // Finalize message and its last part
        const parts = [
            ...lastMessage.parts || []
        ];
        if (parts.length > 0) parts[parts.length - 1] = {
            ...parts[parts.length - 1],
            final: true
        };
        messages[lastMessageIndex] = {
            ...lastMessage,
            parts: parts,
            final: true,
            updatedAt: new Date().toISOString()
        };
        $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageUpdated", messages[lastMessageIndex]);
    }
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$cd74f809c89ea10c(get, set, role) {
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const lastMessageIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === role);
    if (lastMessageIndex === -1) return;
    const lastMessage = messages[lastMessageIndex];
    if ($cbb62fb16c5d45ff$export$f4610a88b89838b(lastMessage)) {
        messages.splice(lastMessageIndex, 1);
        const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
        set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
    }
}
function $cbb62fb16c5d45ff$export$c3a007efc2f315a0(get, set, messageData) {
    const now = new Date();
    const current = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
    // If the most recent message is an active (non-final) assistant message,
    // backdate the injected message so it sorts before the assistant's createdAt.
    // This prevents the injected message from splitting the bot's response across
    // two bubbles and breaking the karaoke cursor.
    const lastMessage = current[current.length - 1];
    const lastAssistant = lastMessage?.role === "assistant" ? lastMessage : undefined;
    let timestamp;
    if (lastAssistant && lastAssistant.final === false && messageData.role === "system") {
        const assistantTime = new Date(lastAssistant.createdAt);
        timestamp = new Date(assistantTime.getTime() - 1).toISOString();
    } else timestamp = now.toISOString();
    const message = {
        role: messageData.role,
        final: true,
        parts: [
            ...messageData.parts
        ],
        createdAt: timestamp,
        updatedAt: now.toISOString()
    };
    const updatedMessages = [
        ...current,
        message
    ];
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(updatedMessages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageCreated", message);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$57de01211e3df548(get, set, text, final) {
    const now = new Date();
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    // Find last user message
    const lastUserIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (m)=>m.role === "user");
    if (lastUserIndex !== -1 && !messages[lastUserIndex].final) {
        // Update existing user message
        const target = {
            ...messages[lastUserIndex]
        };
        const parts = Array.isArray(target.parts) ? [
            ...target.parts
        ] : [];
        const lastPart = parts[parts.length - 1];
        if (!lastPart || lastPart.final) // Start a new part
        parts.push({
            text: text,
            final: final,
            createdAt: now.toISOString()
        });
        else // Update in-progress part
        parts[parts.length - 1] = {
            ...lastPart,
            text: text,
            final: final
        };
        const updatedMessage = {
            ...target,
            parts: parts,
            updatedAt: now.toISOString()
        };
        messages[lastUserIndex] = updatedMessage;
        const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
        $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageUpdated", updatedMessage);
        set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
        return;
    }
    // Create a new user message initialized with this transcript
    const newMessage = {
        role: "user",
        final: false,
        parts: [
            {
                text: text,
                final: final,
                createdAt: now.toISOString()
            }
        ],
        createdAt: now.toISOString(),
        updatedAt: now.toISOString()
    };
    const updatedMessages = [
        ...messages,
        newMessage
    ];
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(updatedMessages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageCreated", newMessage);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$d78810e35f742a16(get, set, text, final, spoken, aggregatedBy) {
    // Strip turn-completion markers emitted by FilterIncompleteTurns before
    // the text enters conversation state.
    text = $cbb62fb16c5d45ff$export$146b18879fa8cfd7(text);
    if (text.length === 0) return;
    const now = new Date();
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const botOutputMessageState = new Map(get((0, $28e6b30207a0392e$export$bd01e11bc95333a1)));
    const lastAssistantIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === "assistant");
    let messageId;
    let isNewMessage = false;
    if (lastAssistantIndex === -1) {
        // Create new assistant message
        isNewMessage = true;
        messageId = now.toISOString();
        const newMessage = {
            role: "assistant",
            final: final,
            parts: [],
            createdAt: messageId,
            updatedAt: messageId
        };
        messages.push(newMessage);
        // Initialize message state
        botOutputMessageState.set(messageId, {
            currentPartIndex: 0,
            currentCharIndex: 0,
            partFinalFlags: [],
            partSpokenOnly: [],
            hasReceivedUnspoken: false
        });
    } else {
        // Update existing assistant message
        const lastMessage = messages[lastAssistantIndex];
        messageId = lastMessage.createdAt;
        messages[lastAssistantIndex] = {
            ...lastMessage,
            final: final ? true : lastMessage.final,
            updatedAt: now.toISOString()
        };
        // Ensure message state exists
        if (!botOutputMessageState.has(messageId)) botOutputMessageState.set(messageId, {
            currentPartIndex: 0,
            currentCharIndex: 0,
            partFinalFlags: [],
            partSpokenOnly: [],
            hasReceivedUnspoken: false
        });
    }
    // Store raw event for debugging/replay
    const botOutputEvents = new Map(get((0, $28e6b30207a0392e$export$3365e5c5967ff793)));
    const existingEvents = botOutputEvents.get(messageId) || [];
    const rawEvent = {
        text: text,
        spoken: spoken,
        aggregatedBy: aggregatedBy,
        receivedAt: now.toISOString()
    };
    botOutputEvents.set(messageId, [
        ...existingEvents,
        rawEvent
    ]);
    const messageState = botOutputMessageState.get(messageId);
    const message = messages[lastAssistantIndex === -1 ? messages.length - 1 : lastAssistantIndex];
    const parts = [
        ...message.parts || []
    ];
    if (!spoken) {
        // UNSPOKEN EVENT: Create/update message parts immediately
        // Only append when both current and last part are word-level; sentence-level
        // and other types each get their own part so spoken events can match 1:1.
        const isDefaultType = aggregatedBy === "sentence" || aggregatedBy === "word" || !aggregatedBy;
        const lastPart = parts[parts.length - 1];
        const shouldAppend = lastPart && aggregatedBy === "word" && lastPart.aggregatedBy === "word" && typeof lastPart.text === "string" && !messageState.partSpokenOnly[parts.length - 1];
        if (shouldAppend) {
            // Append to last part (word-level only)
            const lastPartText = lastPart.text;
            const separator = lastPartText && !lastPartText.endsWith(" ") && !text.startsWith(" ") ? " " : "";
            parts[parts.length - 1] = {
                ...lastPart,
                text: lastPartText + separator + text
            };
        } else {
            // Before creating a new unspoken part, check whether the trailing run
            // of spoken-only fallback parts forms a word-prefix of this new text.
            // If so, absorb them: S2S pipelines can emit word-level spoken events
            // for a sentence before the sentence-level unspoken event arrives.
            let scanStart = parts.length;
            while(scanStart > 0 && messageState.partSpokenOnly[scanStart - 1] && typeof parts[scanStart - 1].text === "string")scanStart--;
            // Character-level prefix match: concatenate the alphanumeric chars
            // of each trailing spoken-only part in order and check that the
            // running concatenation is a prefix of the new text's alphanumeric
            // chars. This handles TTS sub-word splits (e.g. "Pi"/"pec"/"at"
            // speaking "Pipecat") that a word-level matcher can't reconcile.
            let absorbed = 0;
            let absorbedAlnumCount = 0;
            if (scanStart < parts.length) {
                const alnumOnly = (s)=>s.toLowerCase().replace(/[^\p{L}\p{N}]/gu, "");
                const textAlnum = alnumOnly(text);
                let running = 0;
                for(let i = scanStart; i < parts.length; i++){
                    const partAlnum = alnumOnly(parts[i].text);
                    if (partAlnum.length === 0) {
                        // Pure-punctuation fragment: absorb silently without extending
                        // the running prefix (same semantics as applySpokenBotOutputProgress).
                        absorbed++;
                        continue;
                    }
                    const next = running + partAlnum.length;
                    if (next <= textAlnum.length && textAlnum.slice(running, next) === partAlnum) {
                        running = next;
                        absorbedAlnumCount = running;
                        absorbed++;
                    } else break;
                }
            }
            // Translate the absorbed alphanumeric prefix count into a char
            // index inside the original (non-normalized) unspoken text,
            // advancing past any trailing punctuation and whitespace so the
            // cursor sits at the start of the next unspoken word.
            let absorbedCharIndex = 0;
            if (absorbedAlnumCount > 0) {
                const isAlnum = (c)=>/[\p{L}\p{N}]/u.test(c);
                let count = 0;
                for(let i = 0; i < text.length; i++)if (isAlnum(text[i])) {
                    count++;
                    if (count === absorbedAlnumCount) {
                        let j = i + 1;
                        while(j < text.length && !isAlnum(text[j]))j++;
                        absorbedCharIndex = j;
                        break;
                    }
                }
            }
            // Create new part (sentence-level, custom types, or first word chunk)
            // Default to inline; custom types get displayMode from metadata in the hook
            const defaultDisplayMode = isDefaultType ? "inline" : undefined;
            const newPart = {
                text: text,
                final: false,
                createdAt: now.toISOString(),
                aggregatedBy: aggregatedBy,
                displayMode: defaultDisplayMode
            };
            if (absorbed > 0) {
                // Replace the absorbed spoken-only parts in place so the new
                // unspoken part sits where its spoken fragments originally landed
                // (before any trailing fallback parts that belong to later turns).
                const insertAt = scanStart;
                parts.splice(insertAt, absorbed, newPart);
                messageState.partFinalFlags.splice(insertAt, absorbed, false);
                messageState.partSpokenOnly.splice(insertAt, absorbed, false);
                if (absorbedCharIndex > 0) {
                    messageState.currentPartIndex = insertAt;
                    messageState.currentCharIndex = absorbedCharIndex;
                    if (absorbedCharIndex >= text.length) messageState.partFinalFlags[insertAt] = true;
                }
            } else {
                parts.push(newPart);
                messageState.partFinalFlags.push(false);
                messageState.partSpokenOnly.push(false);
            }
        }
        messageState.hasReceivedUnspoken = true;
        // Update message with new parts
        messages[lastAssistantIndex === -1 ? messages.length - 1 : lastAssistantIndex] = {
            ...message,
            parts: parts
        };
    } else {
        // SPOKEN EVENT: advance cursor into existing text, or add as new part
        // if there is none. Multi-word events (e.g. TTS emitting "Hello! I'm"
        // as a single chunk that crosses a sentence boundary) are split into
        // whitespace-separated tokens and processed one at a time so the
        // cursor can advance across unspoken-part boundaries.
        const tokens = text.trim().split(/\s+/).filter(Boolean);
        const leadingSpace = /^\s/.test(text) ? " " : "";
        const isDefaultType = aggregatedBy === "sentence" || aggregatedBy === "word" || !aggregatedBy;
        const defaultDisplayMode = isDefaultType ? "inline" : undefined;
        let partsMutated = false;
        for(let tokenIdx = 0; tokenIdx < tokens.length; tokenIdx++){
            const tokenText = (tokenIdx === 0 ? leadingSpace : " ") + tokens[tokenIdx];
            const advanced = parts.length > 0 && (0, $ad0c8ea1ee293a79$export$9b39fd8438081d6d)(messageState, parts, tokenText);
            if (advanced) continue;
            if (messageState.hasReceivedUnspoken) {
                if ((0, $ad0c8ea1ee293a79$export$802e9b947136afbf)(messageState, parts)) continue;
                if ((0, $ad0c8ea1ee293a79$export$7d7fdcf7579a70b2)(tokenText).trim().length === 0) continue;
            }
            // Either no unspoken content has ever been seen (spoken-only bot),
            // or every part has been fully consumed and this token likely
            // belongs to the next sentence — whose unspoken event hasn't arrived
            // yet. Add as a spoken-only fallback so a later unspoken event can
            // absorb it.
            const newPart = {
                text: tokenText,
                final: false,
                createdAt: now.toISOString(),
                aggregatedBy: aggregatedBy,
                displayMode: defaultDisplayMode
            };
            parts.push(newPart);
            messageState.partFinalFlags.push(true);
            messageState.partSpokenOnly.push(true);
            messageState.currentPartIndex = parts.length - 1;
            messageState.currentCharIndex = tokenText.length;
            partsMutated = true;
        }
        if (partsMutated) messages[lastAssistantIndex === -1 ? messages.length - 1 : lastAssistantIndex] = {
            ...message,
            parts: parts
        };
    }
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
    const cbs = get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab));
    const updatedMsg = messages[lastAssistantIndex === -1 ? messages.length - 1 : lastAssistantIndex];
    if (isNewMessage) $cbb62fb16c5d45ff$var$callCallbacks(cbs, "onMessageCreated", updatedMsg);
    else $cbb62fb16c5d45ff$var$callCallbacks(cbs, "onMessageUpdated", updatedMsg);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
    set((0, $28e6b30207a0392e$export$bd01e11bc95333a1), botOutputMessageState);
    set((0, $28e6b30207a0392e$export$3365e5c5967ff793), botOutputEvents);
}
function $cbb62fb16c5d45ff$export$e450805545c7dd86(get, set, data) {
    const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
    // If a tool_call_id is provided, check for an existing entry to avoid duplicates
    if (data.tool_call_id) {
        const existingIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === "function_call" && msg.functionCall?.tool_call_id === data.tool_call_id);
        if (existingIndex !== -1) return;
    }
    const now = new Date();
    // If the most recent message is an active (non-final) assistant message,
    // backdate the function call so it sorts before the assistant's createdAt.
    // This prevents the function call from splitting the bot's response across
    // two bubbles and breaking the karaoke cursor.
    const lastMessage = messages[messages.length - 1];
    let timestamp;
    if (lastMessage?.role === "assistant" && lastMessage.final === false) {
        const assistantTime = new Date(lastMessage.createdAt);
        timestamp = new Date(assistantTime.getTime() - 1).toISOString();
    } else timestamp = now.toISOString();
    const message = {
        role: "function_call",
        final: false,
        parts: [],
        createdAt: timestamp,
        updatedAt: now.toISOString(),
        functionCall: {
            function_name: data.function_name,
            tool_call_id: data.tool_call_id,
            args: data.args,
            status: "started"
        }
    };
    const updatedMessages = [
        ...messages,
        message
    ];
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(updatedMessages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageCreated", message);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
}
function $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, tool_call_id, updates) {
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const index = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === "function_call" && msg.functionCall?.tool_call_id === tool_call_id);
    if (index === -1) return false;
    const existing = messages[index];
    const updated = {
        ...existing,
        updatedAt: new Date().toISOString(),
        final: updates.status === "completed" ? true : existing.final,
        functionCall: {
            ...existing.functionCall,
            ...updates
        }
    };
    messages[index] = updated;
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageUpdated", updated);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
    return true;
}
function $cbb62fb16c5d45ff$export$54b3765e641b1813(get, set, updates) {
    const messages = [
        ...get((0, $28e6b30207a0392e$export$92a4076839a978d0))
    ];
    const index = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === "function_call" && msg.functionCall?.status === "started" && !msg.functionCall?.tool_call_id);
    if (index === -1) return false;
    const existing = messages[index];
    const updated = {
        ...existing,
        updatedAt: new Date().toISOString(),
        functionCall: {
            ...existing.functionCall,
            ...updates
        }
    };
    messages[index] = updated;
    const processedMessages = $cbb62fb16c5d45ff$var$normalizeMessagesForUI(messages);
    $cbb62fb16c5d45ff$var$callCallbacks(get((0, $28e6b30207a0392e$export$b0cd27059c88d0ab)), "onMessageUpdated", updated);
    set((0, $28e6b30207a0392e$export$92a4076839a978d0), processedMessages);
    return true;
}
function $cbb62fb16c5d45ff$export$809efdca1a10761a(get, set, data) {
    const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
    const lastFc = (0, $f4fa178c5adc67bb$export$296de88ccac4bedb)(messages, (m)=>m.role === "function_call");
    // Check if InProgress already created an entry (events arrived out of order).
    if (lastFc?.functionCall && lastFc.functionCall.status !== "started" && Date.now() - new Date(lastFc.updatedAt ?? lastFc.createdAt).getTime() < 2000) {
        if (data.function_name && !lastFc.functionCall.function_name && lastFc.functionCall.tool_call_id) $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, lastFc.functionCall.tool_call_id, {
            function_name: data.function_name
        });
        return;
    }
    $cbb62fb16c5d45ff$export$e450805545c7dd86(get, set, {
        function_name: data.function_name
    });
}
function $cbb62fb16c5d45ff$export$568cde5f4aa4dcba(get, set, data) {
    // Tier 1: Try to update the last "started" entry (from LLMFunctionCallStarted)
    const updated = $cbb62fb16c5d45ff$export$54b3765e641b1813(get, set, {
        function_name: data.function_name,
        tool_call_id: data.tool_call_id,
        args: data.args,
        status: "in_progress"
    });
    if (!updated) {
        // Tier 2: Try updating an existing entry by tool_call_id
        const found = $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, data.tool_call_id, {
            function_name: data.function_name,
            args: data.args,
            status: "in_progress"
        });
        if (!found) {
            // Tier 3: No existing entry at all; create a new one as in_progress
            $cbb62fb16c5d45ff$export$e450805545c7dd86(get, set, {
                function_name: data.function_name,
                tool_call_id: data.tool_call_id,
                args: data.args
            });
            $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, data.tool_call_id, {
                status: "in_progress"
            });
        }
    }
}
function $cbb62fb16c5d45ff$export$a3cea3f7508e820b(get, set, data) {
    // Tier 1: Try updating by tool_call_id
    const found = $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, data.tool_call_id, {
        function_name: data.function_name,
        status: "completed",
        result: data.result,
        cancelled: data.cancelled
    });
    if (!found) {
        // Tier 2: No match by tool_call_id (e.g. InProgress was skipped).
        const matched = $cbb62fb16c5d45ff$export$54b3765e641b1813(get, set, {
            function_name: data.function_name,
            tool_call_id: data.tool_call_id
        });
        if (matched) $cbb62fb16c5d45ff$export$1908264de3a8afe4(get, set, data.tool_call_id, {
            status: "completed",
            result: data.result,
            cancelled: data.cancelled
        });
    }
}


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ /**
 * Apply a {@link BotOutputFilter} to a {@link BotOutputText}, zeroing out any
 * portion whose flag is explicitly `false`. Both flags default to `true`, so
 * an omitted filter (or `undefined`) returns the text unchanged.
 */ function $62d3a839f4db69ec$export$ea20379ac23e22f2(text, filter) {
    return {
        spoken: filter?.spoken === false ? "" : text.spoken,
        unspoken: filter?.unspoken === false ? "" : text.unspoken
    };
}








/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $a37ab5d1bcf1d3b2$export$fe23d0ef95286467 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)({
    on: ()=>{},
    off: ()=>{}
});


const $8a6b68ebf0332682$export$33a6ac53b8f02625 = (event, handler)=>{
    const { on: on, off: off } = (0, $5Zyvw$react.useContext)((0, $a37ab5d1bcf1d3b2$export$fe23d0ef95286467));
    (0, $5Zyvw$react.useEffect)(()=>{
        on(event, handler);
        return ()=>{
            off(event, handler);
        };
    }, [
        event,
        handler,
        on,
        off
    ]);
};






/**
 * Checks if a version meets a minimum version requirement.
 * Inlined to avoid adding a `semver` dependency.
 */ function $40515932ecd2dc41$var$isMinVersion(currentVersion, minVersion) {
    // Strip pre-release suffix (e.g. "1.1.0-beta.1" -> "1.1.0")
    const parts = currentVersion.split("-")[0].split(".").map(Number);
    for(let i = 0; i < 3; i++){
        if ((parts[i] || 0) > minVersion[i]) return true;
        if ((parts[i] || 0) < minVersion[i]) return false;
    }
    return true; // equal
}
/** Delay (ms) before finalizing the assistant message after bot stops speaking. */ const $40515932ecd2dc41$var$BOT_STOPPED_FINALIZE_DELAY_MS = 2500;
function $40515932ecd2dc41$export$52680e26621bab39() {
    const userStoppedTimeout = (0, $5Zyvw$react.useRef)(undefined);
    const botStoppedSpeakingTimeoutRef = (0, $5Zyvw$react.useRef)(undefined);
    const assistantStreamResetRef = (0, $5Zyvw$react.useRef)(0);
    const botOutputLastChunkRef = (0, $5Zyvw$react.useRef)({
        spoken: "",
        unspoken: ""
    });
    // Clean up pending timeouts on unmount
    (0, $5Zyvw$react.useEffect)(()=>{
        return ()=>{
            clearTimeout(userStoppedTimeout.current);
            clearTimeout(botStoppedSpeakingTimeoutRef.current);
        };
    }, []);
    // -- helpers ---------------------------------------------------------------
    const finalizeLastAssistantMessageIfPending = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        clearTimeout(botStoppedSpeakingTimeoutRef.current);
        botStoppedSpeakingTimeoutRef.current = undefined;
        const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
        const lastAssistant = (0, $f4fa178c5adc67bb$export$296de88ccac4bedb)(messages, (m)=>m.role === "assistant");
        if (lastAssistant && !lastAssistant.final) (0, $cbb62fb16c5d45ff$export$78bd074993ab081f)(get, set, "assistant");
    }, []));
    const ensureAssistantMessage = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
        const lastAssistantIndex = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (msg)=>msg.role === "assistant");
        const lastAssistant = lastAssistantIndex !== -1 ? messages[lastAssistantIndex] : undefined;
        if (!lastAssistant || lastAssistant.final) {
            // If the message was finalized but still has unspoken content, it was
            // finalized prematurely (e.g. BotStoppedSpeaking timer fired during a
            // TTS pause mid-response). Un-finalize it instead of creating a new
            // message bubble — but only when no user message followed.
            if (lastAssistant?.final && lastAssistantIndex === messages.length - 1) {
                const messageId = lastAssistant.createdAt;
                const botOutputState = get((0, $28e6b30207a0392e$export$bd01e11bc95333a1));
                const cursor = botOutputState.get(messageId);
                if (cursor && (0, $ad0c8ea1ee293a79$export$802e9b947136afbf)(cursor, lastAssistant.parts || [])) {
                    (0, $cbb62fb16c5d45ff$export$a05f96a2aa40873e)(get, set, "assistant", {
                        final: false
                    });
                    return false;
                }
            }
            (0, $cbb62fb16c5d45ff$export$16fdb433d434f08)(get, set, {
                role: "assistant",
                final: false,
                parts: []
            });
            assistantStreamResetRef.current += 1;
            return true;
        }
        return false;
    }, []));
    // -- event handlers --------------------------------------------------------
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).Connected, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        (0, $cbb62fb16c5d45ff$export$f2434643f2abff11)(get, set);
        set((0, $28e6b30207a0392e$export$57b1a1df06a7e728), null);
        clearTimeout(botStoppedSpeakingTimeoutRef.current);
        botStoppedSpeakingTimeoutRef.current = undefined;
        botOutputLastChunkRef.current = {
            spoken: "",
            unspoken: ""
        };
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).BotReady, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, botData)=>{
        const rtviVersion = botData.version;
        const supportsBotOutput = $40515932ecd2dc41$var$isMinVersion(rtviVersion, [
            1,
            1,
            0
        ]);
        set((0, $28e6b30207a0392e$export$57b1a1df06a7e728), supportsBotOutput);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).BotOutput, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, data)=>{
        // A BotOutput event means the response is still active; cancel any
        // pending finalize timer from BotStoppedSpeaking.
        clearTimeout(botStoppedSpeakingTimeoutRef.current);
        botStoppedSpeakingTimeoutRef.current = undefined;
        ensureAssistantMessage();
        // Handle spacing for BotOutput chunks
        let textToAdd = data.text;
        const lastChunk = data.spoken ? botOutputLastChunkRef.current.spoken : botOutputLastChunkRef.current.unspoken;
        // Add space separator if needed between BotOutput chunks
        if (lastChunk) textToAdd = " " + textToAdd;
        // Update the appropriate last chunk tracker
        if (data.spoken) botOutputLastChunkRef.current.spoken = textToAdd;
        else botOutputLastChunkRef.current.unspoken = textToAdd;
        // Update both spoken and unspoken text streams
        const isFinal = data.aggregated_by === "sentence";
        (0, $cbb62fb16c5d45ff$export$d78810e35f742a16)(get, set, textToAdd, isFinal, data.spoken, data.aggregated_by);
    }, [
        ensureAssistantMessage
    ])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).BotStoppedSpeaking, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        // Don't finalize immediately; start a timer. Bot may start speaking again (pause).
        clearTimeout(botStoppedSpeakingTimeoutRef.current);
        const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
        const lastAssistant = (0, $f4fa178c5adc67bb$export$296de88ccac4bedb)(messages, (m)=>m.role === "assistant");
        if (!lastAssistant || lastAssistant.final) return;
        botStoppedSpeakingTimeoutRef.current = setTimeout(()=>{
            botStoppedSpeakingTimeoutRef.current = undefined;
            // Snap the speech-progress cursor to the end of all parts.
            // The bot finished speaking normally (not interrupted), so all
            // text should render as "spoken". Without this, text from the
            // last sentence can remain grey if the spoken BotOutput event
            // didn't match the unspoken text exactly.
            const msgs = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
            const cursorMap = new Map(get((0, $28e6b30207a0392e$export$bd01e11bc95333a1)));
            const last = (0, $f4fa178c5adc67bb$export$296de88ccac4bedb)(msgs, (m)=>m.role === "assistant");
            if (last) {
                const cursor = cursorMap.get(last.createdAt);
                if (cursor && last.parts && last.parts.length > 0) {
                    const lastPartIdx = last.parts.length - 1;
                    const lastPartText = last.parts[lastPartIdx]?.text;
                    cursor.currentPartIndex = lastPartIdx;
                    cursor.currentCharIndex = typeof lastPartText === "string" ? lastPartText.length : 0;
                    for(let i = 0; i <= lastPartIdx; i++)cursor.partFinalFlags[i] = true;
                    set((0, $28e6b30207a0392e$export$bd01e11bc95333a1), cursorMap);
                }
            }
            (0, $cbb62fb16c5d45ff$export$78bd074993ab081f)(get, set, "assistant");
        }, $40515932ecd2dc41$var$BOT_STOPPED_FINALIZE_DELAY_MS);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).BotStartedSpeaking, (0, $5Zyvw$react.useCallback)(()=>{
        // Bot is speaking again; reset the finalize timer (bot was just pausing).
        clearTimeout(botStoppedSpeakingTimeoutRef.current);
        botStoppedSpeakingTimeoutRef.current = undefined;
    }, []));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).UserStartedSpeaking, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        // User started a new turn; bot's turn is done. Fast-forward: finalize immediately.
        finalizeLastAssistantMessageIfPending();
        clearTimeout(userStoppedTimeout.current);
        // Only finalize the previous user message if the bot has responded since
        // the user last spoke. This prevents finalizing during VAD gaps (brief
        // breathing pauses within the same user turn where UserStoppedSpeaking/
        // UserStartedSpeaking fire without an actual turn change).
        const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
        const lastUserIdx = (0, $f4fa178c5adc67bb$export$8855a8be7bd3e9f8)(messages, (m)=>m.role === "user");
        if (lastUserIdx !== -1 && !messages[lastUserIdx].final) {
            const hasBotActivityAfterUser = messages.slice(lastUserIdx + 1).some((m)=>m.role === "assistant");
            if (hasBotActivityAfterUser) (0, $cbb62fb16c5d45ff$export$78bd074993ab081f)(get, set, "user");
        }
    }, [
        finalizeLastAssistantMessageIfPending
    ])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).UserTranscript, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, data)=>{
        const text = data.text ?? "";
        const final = Boolean(data.final);
        (0, $cbb62fb16c5d45ff$export$57de01211e3df548)(get, set, text, final);
        // If we got any transcript, cancel pending cleanup
        clearTimeout(userStoppedTimeout.current);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).UserStoppedSpeaking, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        clearTimeout(userStoppedTimeout.current);
        // If no transcript ends up arriving, ensure any accidental empty placeholder is removed.
        userStoppedTimeout.current = setTimeout(()=>{
            // Re-read state at timeout time
            const messages = get((0, $28e6b30207a0392e$export$92a4076839a978d0));
            const lastUser = (0, $f4fa178c5adc67bb$export$296de88ccac4bedb)(messages, (m)=>m.role === "user");
            const hasParts = Array.isArray(lastUser?.parts) && lastUser.parts.length > 0;
            if (!lastUser || !hasParts) (0, $cbb62fb16c5d45ff$export$cd74f809c89ea10c)(get, set, "user");
            else if (!lastUser.final) (0, $cbb62fb16c5d45ff$export$78bd074993ab081f)(get, set, "user");
        }, 3000);
    }, [])));
    // LLM Function Call lifecycle events
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).LLMFunctionCallStarted, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, data)=>{
        (0, $cbb62fb16c5d45ff$export$809efdca1a10761a)(get, set, {
            function_name: data.function_name
        });
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).LLMFunctionCallInProgress, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, data)=>{
        (0, $cbb62fb16c5d45ff$export$568cde5f4aa4dcba)(get, set, {
            function_name: data.function_name,
            tool_call_id: data.tool_call_id,
            args: data.arguments
        });
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).LLMFunctionCallStopped, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, data)=>{
        (0, $cbb62fb16c5d45ff$export$a3cea3f7508e820b)(get, set, {
            function_name: data.function_name,
            tool_call_id: data.tool_call_id,
            result: data.result,
            cancelled: data.cancelled
        });
    }, [])));
}


const $984246ab4e966dee$export$4d68d6035e4164b1 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)(null);
const $984246ab4e966dee$export$e372a2172e7d18e8 = ({ children: children })=>{
    (0, $40515932ecd2dc41$export$52680e26621bab39)();
    const injectMessage = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, message)=>{
        (0, $cbb62fb16c5d45ff$export$c3a007efc2f315a0)(get, set, message);
    }, []));
    const botOutputSupported = (0, $5Zyvw$jotai.useAtomValue)((0, $28e6b30207a0392e$export$57b1a1df06a7e728));
    return (0, $5Zyvw$reactjsxruntime.jsx)($984246ab4e966dee$export$4d68d6035e4164b1.Provider, {
        value: {
            injectMessage: injectMessage,
            botOutputSupported: botOutputSupported
        },
        children: children
    });
};
$984246ab4e966dee$export$e372a2172e7d18e8.displayName = "PipecatConversationProvider";
const $984246ab4e966dee$export$8eec679bf8249822 = ()=>{
    const context = (0, $5Zyvw$react.useContext)($984246ab4e966dee$export$4d68d6035e4164b1);
    if (!context) throw new Error("useConversationContext must be used within a PipecatClientProvider");
    return context;
};





/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 



/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 





var $d1f89db81f967f39$exports = {};
$d1f89db81f967f39$exports = JSON.parse("{\"name\":\"@pipecat-ai/client-react\",\"version\":\"1.3.0\",\"license\":\"BSD-2-Clause\",\"main\":\"dist/index.js\",\"module\":\"dist/index.module.js\",\"types\":\"dist/index.d.ts\",\"source\":\"src/index.ts\",\"repository\":{\"type\":\"git\",\"url\":\"git+https://github.com/pipecat-ai/pipecat-client-web.git\"},\"exports\":{\".\":{\"types\":\"./dist/index.d.ts\",\"import\":\"./dist/index.module.js\",\"require\":\"./dist/index.js\"}},\"files\":[\"dist\",\"package.json\",\"README.md\"],\"scripts\":{\"build\":\"parcel build --no-cache\",\"dev\":\"parcel watch\",\"lint\":\"eslint . --report-unused-disable-directives --max-warnings 0 --ignore-pattern 'dist/'\",\"test\":\"jest\"},\"jest\":{\"preset\":\"ts-jest\",\"testEnvironment\":\"jsdom\",\"moduleNameMapper\":{\"^@/(.*)$\":\"<rootDir>/src/$1\"},\"transform\":{\"^.+\\\\.tsx?$\":[\"ts-jest\",{\"tsconfig\":{\"jsx\":\"react-jsx\",\"module\":\"commonjs\",\"moduleResolution\":\"node\",\"esModuleInterop\":true,\"allowImportingTsExtensions\":false,\"noUnusedLocals\":false,\"noUnusedParameters\":false}}]}},\"devDependencies\":{\"@jest/globals\":\"^30.2.0\",\"@pipecat-ai/client-js\":\"*\",\"@types/jest\":\"^30.0.0\",\"@types/react\":\"^18.3.3\",\"@types/react-dom\":\"^18.3.0\",\"@typescript-eslint/eslint-plugin\":\"^8.32.0\",\"eslint\":\"^9.11.1\",\"eslint-config-prettier\":\"^9.1.0\",\"eslint-plugin-react-hooks\":\"^5.2.0\",\"eslint-plugin-simple-import-sort\":\"^12.1.1\",\"jest\":\"^30.2.0\",\"jest-environment-jsdom\":\"^30.2.0\",\"parcel\":\"^2.12.0\",\"react\":\"^18.3.1\",\"react-dom\":\"^18.3.1\",\"ts-jest\":\"^29.4.6\",\"typescript\":\"^5.2.2\"},\"peerDependencies\":{\"@pipecat-ai/client-js\":\"*\",\"react\":\">=18\",\"react-dom\":\">=18\"},\"dependencies\":{\"jotai\":\"^2.9.0\"}}");








const $44c6278d186dd94f$export$d6bdcccacef16204 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)({
    enableCam: ()=>{
        throw new Error("PipecatClientCamStateContext: enableCam() called outside of provider");
    },
    isCamEnabled: false
});
const $44c6278d186dd94f$export$802b42df0e0d8153 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)({
    enableMic: ()=>{
        throw new Error("PipecatClientMicStateContext: enableMic() called outside of provider");
    },
    isMicEnabled: false
});
const $44c6278d186dd94f$export$8e633e67c760098b = /*#__PURE__*/ (0, $5Zyvw$react.createContext)({
    enableScreenShare: ()=>{
        throw new Error("PipecatClientScreenShareStateContext: enableScreenShare() called outside of provider");
    },
    isScreenShareEnabled: false
});
const $44c6278d186dd94f$export$db79fdf85ddd6b65 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)("disconnected");
const $44c6278d186dd94f$export$4777554fda61c378 = ({ children: children })=>{
    const client = (0, $172f489fc5d91d99$export$777fa8498be78705)();
    const [isCamEnabled, setIsCamEnabled] = (0, $5Zyvw$react.useState)(false);
    const [isMicEnabled, setIsMicEnabled] = (0, $5Zyvw$react.useState)(false);
    const [isScreenShareEnabled, setIsScreenShareEnabled] = (0, $5Zyvw$react.useState)(false);
    const [transportState, setTransportState] = (0, $5Zyvw$react.useState)("disconnected");
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).TransportStateChanged, (state)=>{
        setTransportState(state);
        if (state === "initialized" && client) {
            setIsCamEnabled(client.isCamEnabled ?? false);
            setIsMicEnabled(client.isMicEnabled ?? false);
            setIsScreenShareEnabled(client.isSharingScreen ?? false);
        }
    });
    const enableCam = (0, $5Zyvw$react.useCallback)((enabled)=>{
        setIsCamEnabled(enabled);
        client?.enableCam?.(enabled);
    }, [
        client
    ]);
    const enableMic = (0, $5Zyvw$react.useCallback)((enabled)=>{
        setIsMicEnabled(enabled);
        client?.enableMic?.(enabled);
    }, [
        client
    ]);
    const enableScreenShare = (0, $5Zyvw$react.useCallback)((enabled)=>{
        client?.enableScreenShare?.(enabled);
    }, [
        client
    ]);
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).ScreenTrackStarted, (_track, participant)=>{
        if (participant?.local) setIsScreenShareEnabled(true);
    });
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).ScreenTrackStopped, (_track, participant)=>{
        if (participant?.local) setIsScreenShareEnabled(false);
    });
    return (0, $5Zyvw$reactjsxruntime.jsx)($44c6278d186dd94f$export$db79fdf85ddd6b65.Provider, {
        value: transportState,
        children: (0, $5Zyvw$reactjsxruntime.jsx)($44c6278d186dd94f$export$d6bdcccacef16204.Provider, {
            value: {
                enableCam: enableCam,
                isCamEnabled: isCamEnabled
            },
            children: (0, $5Zyvw$reactjsxruntime.jsx)($44c6278d186dd94f$export$802b42df0e0d8153.Provider, {
                value: {
                    enableMic: enableMic,
                    isMicEnabled: isMicEnabled
                },
                children: (0, $5Zyvw$reactjsxruntime.jsx)($44c6278d186dd94f$export$8e633e67c760098b.Provider, {
                    value: {
                        enableScreenShare: enableScreenShare,
                        isScreenShareEnabled: isScreenShareEnabled
                    },
                    children: children
                })
            })
        })
    });
};



const $8df0e777e4d7dd49$var$defaultStore = (0, $5Zyvw$jotai.createStore)();
const $8df0e777e4d7dd49$export$67f6d73bc6cd7bb1 = /*#__PURE__*/ (0, $5Zyvw$react.createContext)({});
const $8df0e777e4d7dd49$export$bb43666ced7a20d0 = ({ children: children, client: client, jotaiStore: jotaiStore = $8df0e777e4d7dd49$var$defaultStore })=>{
    (0, $5Zyvw$react.useEffect)(()=>{
        (0, $5Zyvw$pipecataiclientjs.setAboutClient)({
            library: (0, $d1f89db81f967f39$exports.name),
            library_version: (0, $d1f89db81f967f39$exports.version)
        });
    }, []);
    const eventHandlersMap = (0, $5Zyvw$react.useRef)({});
    (0, $5Zyvw$react.useEffect)(()=>{
        if (!client) return;
        const allEvents = Object.values((0, $5Zyvw$pipecataiclientjs.RTVIEvent)).filter((value)=>isNaN(Number(value)));
        const allHandlers = {};
        allEvents.forEach((event)=>{
            const handler = (...payload)=>{
                const handlers = eventHandlersMap.current[event];
                if (!handlers) return;
                handlers.forEach((h)=>{
                    h(...payload);
                });
            };
            allHandlers[event] = handler;
            client.on(event, handler);
        });
        return ()=>{
            allEvents.forEach((event)=>{
                client.off(event, allHandlers[event]);
            });
        };
    }, [
        client
    ]);
    const on = (0, $5Zyvw$react.useCallback)((event, handler)=>{
        if (!eventHandlersMap.current[event]) eventHandlersMap.current[event] = new Set();
        eventHandlersMap.current[event].add(handler);
    }, []);
    const off = (0, $5Zyvw$react.useCallback)((event, handler)=>{
        eventHandlersMap.current[event]?.delete(handler);
    }, []);
    return (0, $5Zyvw$reactjsxruntime.jsx)((0, $5Zyvw$jotaireact.Provider), {
        store: jotaiStore,
        children: (0, $5Zyvw$reactjsxruntime.jsx)($8df0e777e4d7dd49$export$67f6d73bc6cd7bb1.Provider, {
            value: {
                client: client
            },
            children: (0, $5Zyvw$reactjsxruntime.jsx)((0, $a37ab5d1bcf1d3b2$export$fe23d0ef95286467).Provider, {
                value: {
                    on: on,
                    off: off
                },
                children: (0, $5Zyvw$reactjsxruntime.jsx)((0, $44c6278d186dd94f$export$4777554fda61c378), {
                    children: (0, $5Zyvw$reactjsxruntime.jsx)((0, $984246ab4e966dee$export$e372a2172e7d18e8), {
                        children: children
                    })
                })
            })
        })
    });
};
$8df0e777e4d7dd49$export$bb43666ced7a20d0.displayName = "PipecatClientProvider";


const $172f489fc5d91d99$export$777fa8498be78705 = ()=>{
    const { client: client } = (0, $5Zyvw$react.useContext)((0, $8df0e777e4d7dd49$export$67f6d73bc6cd7bb1));
    return client;
};



const $630203d8dad1dd45$var$localAudioTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$localVideoTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$localScreenAudioTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$localScreenVideoTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$botAudioTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$botVideoTrackAtom = (0, $5Zyvw$jotai.atom)(null);
const $630203d8dad1dd45$var$trackAtom = (0, $5Zyvw$jotaiutils.atomFamily)(({ local: local, trackType: trackType })=>{
    if (local) switch(trackType){
        case "audio":
            return $630203d8dad1dd45$var$localAudioTrackAtom;
        case "screenAudio":
            return $630203d8dad1dd45$var$localScreenAudioTrackAtom;
        case "screenVideo":
            return $630203d8dad1dd45$var$localScreenVideoTrackAtom;
        case "video":
            return $630203d8dad1dd45$var$localVideoTrackAtom;
    }
    return trackType === "audio" ? $630203d8dad1dd45$var$botAudioTrackAtom : $630203d8dad1dd45$var$botVideoTrackAtom;
});
const $630203d8dad1dd45$export$9813dcd2d0c26814 = (trackType, participantType)=>{
    const client = (0, $172f489fc5d91d99$export$777fa8498be78705)();
    const track = (0, $5Zyvw$jotai.useAtomValue)($630203d8dad1dd45$var$trackAtom({
        local: participantType === "local",
        trackType: trackType
    }));
    const updateTrack = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set, track, trackType, local)=>{
        const atom = $630203d8dad1dd45$var$trackAtom({
            local: local,
            trackType: trackType
        });
        const oldTrack = get(atom);
        if (oldTrack?.id === track.id) return;
        set(atom, track);
    }, []));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).TrackStarted, (0, $5Zyvw$react.useCallback)((track, participant)=>{
        updateTrack(track, track.kind, Boolean(participant?.local));
    }, [
        updateTrack
    ]));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).ScreenTrackStarted, (0, $5Zyvw$react.useCallback)((track, participant)=>{
        const trackType = track.kind === "audio" ? "screenAudio" : "screenVideo";
        updateTrack(track, trackType, Boolean(participant?.local));
    }, [
        updateTrack
    ]));
    (0, $5Zyvw$react.useEffect)(()=>{
        if (!client) return;
        const tracks = client.tracks();
        const track = tracks?.[participantType]?.[trackType];
        if (!track) return;
        updateTrack(track, trackType, participantType === "local");
    }, [
        participantType,
        trackType,
        updateTrack,
        client
    ]);
    return track;
};



const $a2a1c5f475114d56$export$b52250cb73ff4de1 = ()=>{
    const botAudioRef = (0, $5Zyvw$react.useRef)(null);
    const botAudioTrack = (0, $630203d8dad1dd45$export$9813dcd2d0c26814)("audio", "bot");
    (0, $5Zyvw$react.useEffect)(()=>{
        if (!botAudioRef.current || !botAudioTrack) return;
        if (botAudioRef.current.srcObject) {
            const oldTrack = botAudioRef.current.srcObject.getAudioTracks()[0];
            if (oldTrack.id === botAudioTrack.id) return;
        }
        botAudioRef.current.srcObject = new MediaStream([
            botAudioTrack
        ]);
    }, [
        botAudioTrack
    ]);
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).SpeakerUpdated, (0, $5Zyvw$react.useCallback)((speaker)=>{
        if (!botAudioRef.current) return;
        if (typeof botAudioRef.current.setSinkId !== "function") return;
        botAudioRef.current.setSinkId(speaker.deviceId);
    }, []));
    return (0, $5Zyvw$reactjsxruntime.jsx)((0, $5Zyvw$reactjsxruntime.Fragment), {
        children: (0, $5Zyvw$reactjsxruntime.jsx)("audio", {
            ref: botAudioRef,
            autoPlay: true
        })
    });
};
$a2a1c5f475114d56$export$b52250cb73ff4de1.displayName = "PipecatClientAudio";




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $d9b24817de62910a$export$3ea2601427f0430f = ()=>(0, $5Zyvw$react.useContext)((0, $44c6278d186dd94f$export$d6bdcccacef16204));


const $666ceba11158b494$export$dc9a029eeca8213f = ({ onCamEnabledChanged: onCamEnabledChanged, disabled: disabled = false, children: children })=>{
    const { isCamEnabled: isCamEnabled, enableCam: enableCam } = (0, $d9b24817de62910a$export$3ea2601427f0430f)();
    const handleToggleCam = (0, $5Zyvw$react.useCallback)(()=>{
        if (disabled) return;
        const newEnabledState = !isCamEnabled;
        enableCam(newEnabledState);
        onCamEnabledChanged?.(newEnabledState);
    }, [
        disabled,
        enableCam,
        isCamEnabled,
        onCamEnabledChanged
    ]);
    return (0, $5Zyvw$reactjsxruntime.jsx)((0, $5Zyvw$reactjsxruntime.Fragment), {
        children: children({
            isCamEnabled: isCamEnabled,
            onClick: handleToggleCam,
            disabled: disabled
        })
    });
};
var $666ceba11158b494$export$2e2bcd8739ae039 = $666ceba11158b494$export$dc9a029eeca8213f;




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $4af0eba414c586fd$export$388e706586309ef0 = ()=>(0, $5Zyvw$react.useContext)((0, $44c6278d186dd94f$export$802b42df0e0d8153));


const $1a9931980e271aa3$export$bc8133b69ff660a2 = ({ onMicEnabledChanged: onMicEnabledChanged, disabled: disabled = false, children: children })=>{
    const { enableMic: enableMic, isMicEnabled: isMicEnabled } = (0, $4af0eba414c586fd$export$388e706586309ef0)();
    const handleToggleMic = (0, $5Zyvw$react.useCallback)(()=>{
        if (disabled) return;
        const newEnabledState = !isMicEnabled;
        enableMic(newEnabledState);
        onMicEnabledChanged?.(newEnabledState);
    }, [
        disabled,
        enableMic,
        isMicEnabled,
        onMicEnabledChanged
    ]);
    return (0, $5Zyvw$reactjsxruntime.jsx)((0, $5Zyvw$reactjsxruntime.Fragment), {
        children: children({
            isMicEnabled: isMicEnabled,
            onClick: handleToggleMic,
            disabled: disabled
        })
    });
};
var $1a9931980e271aa3$export$2e2bcd8739ae039 = $1a9931980e271aa3$export$bc8133b69ff660a2;





/**
 * Copyright (c) 2025, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $5fc67d7ca05dec34$export$be63b19bd7f7d4f5 = ()=>(0, $5Zyvw$react.useContext)((0, $44c6278d186dd94f$export$8e633e67c760098b));


const $0a9d1dc5be2e1cc2$export$93764714dfab3a46 = ({ onScreenShareEnabledChanged: onScreenShareEnabledChanged, disabled: disabled = false, children: children })=>{
    const { enableScreenShare: enableScreenShare, isScreenShareEnabled: isScreenShareEnabled } = (0, $5fc67d7ca05dec34$export$be63b19bd7f7d4f5)();
    const handleToggleScreenShare = (0, $5Zyvw$react.useCallback)(()=>{
        if (disabled) return;
        enableScreenShare(!isScreenShareEnabled);
    }, [
        disabled,
        enableScreenShare,
        isScreenShareEnabled
    ]);
    (0, $5Zyvw$react.useEffect)(()=>{
        onScreenShareEnabledChanged?.(isScreenShareEnabled);
    }, [
        isScreenShareEnabled,
        onScreenShareEnabledChanged
    ]);
    return (0, $5Zyvw$reactjsxruntime.jsx)((0, $5Zyvw$reactjsxruntime.Fragment), {
        children: children({
            isScreenShareEnabled: isScreenShareEnabled,
            onClick: handleToggleScreenShare,
            disabled: disabled
        })
    });
};
var $0a9d1dc5be2e1cc2$export$2e2bcd8739ae039 = $0a9d1dc5be2e1cc2$export$93764714dfab3a46;




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * This file contains code derived from:
 * https://github.com/jaredLunde/react-hook/blob/master/packages/merged-ref/src/index.tsx
 * Original author: Jared Lunde (https://github.com/jaredLunde)
 * Original license: MIT (https://github.com/jaredLunde/react-hook/blob/master/LICENSE)
 */ 
function $a73c18d97634a8ff$var$useMergedRef(...refs) {
    return (0, $5Zyvw$react.useCallback)((element)=>{
        for(let i = 0; i < refs.length; i++){
            const ref = refs[i];
            if (typeof ref === "function") ref(element);
            else if (ref && typeof ref === "object") ref.current = element;
        }
    }, // eslint-disable-next-line react-hooks/exhaustive-deps
    refs);
}
var $a73c18d97634a8ff$export$2e2bcd8739ae039 = $a73c18d97634a8ff$var$useMergedRef;



const $0f97689637ada1d8$export$85974db6d0cc43b3 = /*#__PURE__*/ (0, $5Zyvw$react.forwardRef)(function VoiceClientVideo({ participant: participant = "local", fit: fit = "contain", mirror: mirror, onResize: onResize, style: style = {}, trackType: trackType = "video", ...props }, ref) {
    const videoTrack = (0, $630203d8dad1dd45$export$9813dcd2d0c26814)(trackType, participant);
    const videoEl = (0, $5Zyvw$react.useRef)(null);
    const videoRef = (0, $a73c18d97634a8ff$export$2e2bcd8739ae039)(videoEl, ref);
    /**
     * Handle canplay & picture-in-picture events.
     */ (0, $5Zyvw$react.useEffect)(function setupVideoEvents() {
        const video = videoEl.current;
        if (!video) return;
        const playVideo = ()=>{
            const promise = video.play();
            if (promise !== undefined) promise.then(()=>{
                // All good, playback started.
                video.controls = false;
            }).catch((error)=>{
                // Auto-play was prevented. Show video controls, so user can play video manually.
                video.controls = true;
                console.warn("Failed to play video", error);
            });
        };
        const handleCanPlay = ()=>{
            if (!video.paused) return;
            playVideo();
        };
        const handleEnterPIP = ()=>{
            video.style.transform = "scale(1)";
        };
        const handleLeavePIP = ()=>{
            video.style.transform = "";
            setTimeout(()=>{
                if (video.paused) playVideo();
            }, 100);
        };
        const handleVisibilityChange = ()=>{
            if (document.visibilityState === "hidden") return;
            if (!video.paused) return;
            playVideo();
        };
        video.addEventListener("canplay", handleCanPlay);
        video.addEventListener("enterpictureinpicture", handleEnterPIP);
        video.addEventListener("leavepictureinpicture", handleLeavePIP);
        // Videos can be paused if media was played in another app on iOS.
        document.addEventListener("visibilitychange", handleVisibilityChange);
        return ()=>{
            video.removeEventListener("canplay", handleCanPlay);
            video.removeEventListener("enterpictureinpicture", handleEnterPIP);
            video.removeEventListener("leavepictureinpicture", handleLeavePIP);
            document.removeEventListener("visibilitychange", handleVisibilityChange);
        };
    }, []);
    /**
     * Update srcObject.
     */ (0, $5Zyvw$react.useEffect)(function updateSrcObject() {
        const video = videoEl.current;
        if (!video || !videoTrack) return;
        video.srcObject = new MediaStream([
            videoTrack
        ]);
        video.load();
        return ()=>{
            // clean up when unmounted
            video.srcObject = null;
            video.load();
        };
    }, [
        videoTrack,
        videoTrack?.id
    ]);
    /**
     * Add optional event listener for resize event so the parent component
     * can know the video's native aspect ratio.
     */ (0, $5Zyvw$react.useEffect)(function reportVideoDimensions() {
        const video = videoEl.current;
        if (!onResize || !video) return;
        let frame;
        function handleResize() {
            if (frame) cancelAnimationFrame(frame);
            frame = requestAnimationFrame(()=>{
                const video = videoEl.current;
                if (!video || document.hidden) return;
                const videoWidth = video.videoWidth;
                const videoHeight = video.videoHeight;
                if (videoWidth && videoHeight) onResize?.({
                    aspectRatio: videoWidth / videoHeight,
                    height: videoHeight,
                    width: videoWidth
                });
            });
        }
        handleResize();
        video.addEventListener("loadedmetadata", handleResize);
        video.addEventListener("resize", handleResize);
        return ()=>{
            if (frame) cancelAnimationFrame(frame);
            video.removeEventListener("loadedmetadata", handleResize);
            video.removeEventListener("resize", handleResize);
        };
    }, [
        onResize
    ]);
    return (0, $5Zyvw$reactjsxruntime.jsx)("video", {
        autoPlay: true,
        muted: true,
        playsInline: true,
        ref: videoRef,
        style: {
            objectFit: fit,
            transform: mirror ? "scale(-1, 1)" : "",
            ...style
        },
        ...props
    });
});
$0f97689637ada1d8$export$85974db6d0cc43b3.displayName = "PipecatClientVideo";




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 





const $9bd3e7d3a9d7acd1$var$availableMicsAtom = (0, $5Zyvw$jotai.atom)([]);
const $9bd3e7d3a9d7acd1$var$availableCamsAtom = (0, $5Zyvw$jotai.atom)([]);
const $9bd3e7d3a9d7acd1$var$availableSpeakersAtom = (0, $5Zyvw$jotai.atom)([]);
const $9bd3e7d3a9d7acd1$var$selectedMicAtom = (0, $5Zyvw$jotai.atom)({});
const $9bd3e7d3a9d7acd1$var$selectedCamAtom = (0, $5Zyvw$jotai.atom)({});
const $9bd3e7d3a9d7acd1$var$selectedSpeakerAtom = (0, $5Zyvw$jotai.atom)({});
const $9bd3e7d3a9d7acd1$export$642bc4d2d2a376f1 = ()=>{
    const client = (0, $172f489fc5d91d99$export$777fa8498be78705)();
    const availableCams = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$availableCamsAtom);
    const availableMics = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$availableMicsAtom);
    const availableSpeakers = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$availableSpeakersAtom);
    const selectedCam = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$selectedCamAtom);
    const selectedMic = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$selectedMicAtom);
    const selectedSpeaker = (0, $5Zyvw$jotai.useAtomValue)($9bd3e7d3a9d7acd1$var$selectedSpeakerAtom);
    const initDevices = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)(async (_get, set)=>{
        if (!client) return;
        const availableCams = await client.getAllCams();
        const availableMics = await client.getAllMics();
        const availableSpeakers = await client.getAllSpeakers();
        set($9bd3e7d3a9d7acd1$var$availableCamsAtom, availableCams);
        set($9bd3e7d3a9d7acd1$var$availableMicsAtom, availableMics);
        set($9bd3e7d3a9d7acd1$var$availableSpeakersAtom, availableSpeakers);
        set($9bd3e7d3a9d7acd1$var$selectedCamAtom, client.selectedCam);
        set($9bd3e7d3a9d7acd1$var$selectedMicAtom, client.selectedMic);
        set($9bd3e7d3a9d7acd1$var$selectedSpeakerAtom, client.selectedSpeaker);
    }, [
        client
    ]));
    (0, $5Zyvw$react.useEffect)(()=>{
        initDevices();
    }, [
        initDevices
    ]);
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).AvailableCamsUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, cams)=>{
        set($9bd3e7d3a9d7acd1$var$availableCamsAtom, cams);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).AvailableMicsUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, mics)=>{
        set($9bd3e7d3a9d7acd1$var$availableMicsAtom, mics);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).AvailableSpeakersUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, speakers)=>{
        set($9bd3e7d3a9d7acd1$var$availableSpeakersAtom, speakers);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).CamUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, cam)=>{
        set($9bd3e7d3a9d7acd1$var$selectedCamAtom, cam);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).MicUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, mic)=>{
        set($9bd3e7d3a9d7acd1$var$selectedMicAtom, mic);
    }, [])));
    (0, $8a6b68ebf0332682$export$33a6ac53b8f02625)((0, $5Zyvw$pipecataiclientjs.RTVIEvent).SpeakerUpdated, (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((_get, set, speaker)=>{
        set($9bd3e7d3a9d7acd1$var$selectedSpeakerAtom, speaker);
    }, [])));
    const updateCam = (0, $5Zyvw$react.useCallback)((id)=>{
        client?.updateCam(id);
    }, [
        client
    ]);
    const updateMic = (0, $5Zyvw$react.useCallback)((id)=>{
        client?.updateMic(id);
    }, [
        client
    ]);
    const updateSpeaker = (0, $5Zyvw$react.useCallback)((id)=>{
        client?.updateSpeaker(id);
    }, [
        client
    ]);
    return {
        availableCams: availableCams,
        availableMics: availableMics,
        availableSpeakers: availableSpeakers,
        selectedCam: selectedCam,
        selectedMic: selectedMic,
        selectedSpeaker: selectedSpeaker,
        updateCam: updateCam,
        updateMic: updateMic,
        updateSpeaker: updateSpeaker
    };
};





/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $810478f6ae107062$export$30aee278309a867b = ()=>(0, $5Zyvw$react.useContext)((0, $44c6278d186dd94f$export$db79fdf85ddd6b65));


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 






const $fe6dfdd0b8560c78$export$acb832b064cb6c09 = ({ onMessageCreated: onMessageCreated, onMessageUpdated: onMessageUpdated, onMessageAdded: onMessageAdded, aggregationMetadata: aggregationMetadata, botOutputFilter: botOutputFilter } = {})=>{
    const { injectMessage: injectMessage } = (0, $984246ab4e966dee$export$8eec679bf8249822)();
    // Generate a unique ID for this hook instance
    const callbackId = (0, $5Zyvw$react.useId)();
    // Resolve deprecated onMessageAdded → onMessageCreated
    const resolvedCreated = onMessageCreated ?? onMessageAdded;
    // Register and unregister the callbacks
    const doRegister = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        (0, $cbb62fb16c5d45ff$export$f2820c5e040afe17)(get, set, callbackId, {
            onMessageCreated: resolvedCreated,
            onMessageUpdated: onMessageUpdated
        });
    }, [
        callbackId,
        resolvedCreated,
        onMessageUpdated
    ]));
    const doUnregister = (0, $5Zyvw$jotaiutils.useAtomCallback)((0, $5Zyvw$react.useCallback)((get, set)=>{
        (0, $cbb62fb16c5d45ff$export$46ae1ba121fdc956)(get, set, callbackId);
    }, [
        callbackId
    ]));
    (0, $5Zyvw$react.useEffect)(()=>{
        doRegister();
        return ()=>{
            doUnregister();
        };
    }, [
        doRegister,
        doUnregister
    ]);
    // Get the raw state from atoms
    const messages = (0, $5Zyvw$jotai.useAtomValue)((0, $28e6b30207a0392e$export$92a4076839a978d0));
    const botOutputMessageState = (0, $5Zyvw$jotai.useAtomValue)((0, $28e6b30207a0392e$export$bd01e11bc95333a1));
    // Memoize the filtered messages to prevent infinite loops
    const filteredMessages = (0, $5Zyvw$react.useMemo)(()=>{
        const getMetadata = (part)=>{
            return part.aggregatedBy ? aggregationMetadata?.[part.aggregatedBy] : undefined;
        };
        // Process messages: convert string parts to BotOutputText based on position state
        const processedMessages = messages.map((message)=>{
            if (message.role === "assistant") {
                const messageId = message.createdAt;
                const messageState = botOutputMessageState.get(messageId);
                if (!messageState) // No state yet, return message as-is
                return message;
                const parts = message.parts || [];
                // Find the actual current part index (skip parts that aren't meant to be spoken)
                let actualCurrentPartIndex = messageState.currentPartIndex;
                while(actualCurrentPartIndex < parts.length){
                    const part = parts[actualCurrentPartIndex];
                    if (typeof part?.text !== "string") break;
                    const isSpoken = getMetadata(part)?.isSpoken !== false;
                    if (isSpoken) break;
                    actualCurrentPartIndex++;
                }
                if (parts.length > 0 && actualCurrentPartIndex >= parts.length) actualCurrentPartIndex = parts.length - 1;
                // Convert parts to BotOutputText format based on position state
                const processedParts = parts.map((part, partIndex)=>{
                    // If part text is not a string, it's already processed (e.g., ReactNode)
                    if (typeof part.text !== "string") return part;
                    const metadata = getMetadata(part);
                    const displayMode = part.displayMode ?? metadata?.displayMode ?? "inline";
                    const isSpoken = metadata?.isSpoken !== false;
                    const partText = displayMode === "block" && !isSpoken ? part.text.trim() : part.text;
                    if (!isSpoken) // Non-spoken aggregation types: all content is "unspoken".
                    return {
                        ...part,
                        displayMode: displayMode,
                        text: (0, $62d3a839f4db69ec$export$ea20379ac23e22f2)({
                            spoken: "",
                            unspoken: partText
                        }, botOutputFilter)
                    };
                    // Use cursor split for the part at actualCurrentPartIndex for every message,
                    // so previous (e.g. interrupted) messages keep partially spoken state.
                    const isPartAtCursor = partIndex === actualCurrentPartIndex;
                    const currentCharIndex = messageState.currentCharIndex;
                    const spokenText = isPartAtCursor ? partText.slice(0, currentCharIndex) : partIndex < actualCurrentPartIndex ? partText : "";
                    const unspokenText = isPartAtCursor ? partText.slice(currentCharIndex) : partIndex < actualCurrentPartIndex ? "" : partText;
                    return {
                        ...part,
                        displayMode: displayMode,
                        text: (0, $62d3a839f4db69ec$export$ea20379ac23e22f2)({
                            spoken: spokenText,
                            unspoken: unspokenText
                        }, botOutputFilter)
                    };
                });
                return {
                    ...message,
                    parts: processedParts
                };
            }
            return message;
        });
        // Messages are already normalized (sorted, filtered, deduped, merged) on write.
        return processedMessages;
    }, [
        messages,
        botOutputMessageState,
        aggregationMetadata,
        botOutputFilter
    ]);
    const botOutputEvents = (0, $5Zyvw$jotai.useAtomValue)((0, $28e6b30207a0392e$export$3365e5c5967ff793));
    return {
        messages: filteredMessages,
        injectMessage: injectMessage,
        botOutputEvents: botOutputEvents
    };
};






const $a1dfa75b13e6bb9b$export$59bf27bd43679db6 = /*#__PURE__*/ (0, ($parcel$interopDefault($5Zyvw$react))).memo(({ backgroundColor: backgroundColor = "transparent", barColor: barColor = "black", barCount: barCount = 5, barGap: barGap = 12, barLineCap: barLineCap = "round", barMaxHeight: barMaxHeight = 120, barOrigin: barOrigin = "center", barWidth: barWidth = 30, participantType: participantType })=>{
    const canvasRef = (0, $5Zyvw$react.useRef)(null);
    const track = (0, $630203d8dad1dd45$export$9813dcd2d0c26814)("audio", participantType);
    (0, $5Zyvw$react.useEffect)(()=>{
        if (!canvasRef.current) return;
        const canvasWidth = barCount * barWidth + (barCount - 1) * barGap;
        const canvasHeight = barMaxHeight;
        const canvas = canvasRef.current;
        const scaleFactor = 2;
        // Make canvas fill the width and height of its container
        const resizeCanvas = ()=>{
            canvas.width = canvasWidth * scaleFactor;
            canvas.height = canvasHeight * scaleFactor;
            canvas.style.width = `${canvasWidth}px`;
            canvas.style.height = `${canvasHeight}px`;
            canvasCtx.lineCap = barLineCap;
            canvasCtx.scale(scaleFactor, scaleFactor);
        };
        const canvasCtx = canvas.getContext("2d");
        resizeCanvas();
        if (!track) return;
        const audioContext = new AudioContext();
        const source = audioContext.createMediaStreamSource(new MediaStream([
            track
        ]));
        const analyser = audioContext.createAnalyser();
        analyser.fftSize = 1024;
        source.connect(analyser);
        const frequencyData = new Uint8Array(analyser.frequencyBinCount);
        canvasCtx.lineCap = barLineCap;
        // Create frequency bands based on barCount
        const bands = Array.from({
            length: barCount
        }, (_, i)=>{
            // Use improved logarithmic scale for better frequency distribution
            const minFreq = barCount > 20 ? 200 : 80; // Adjust min frequency based on bar count
            const maxFreq = 10000; // Cover most important audio frequencies
            // Use Mel scale inspired approach for more perceptually uniform distribution
            // This helps with a large number of bars by placing fewer in the very low range
            // https://en.wikipedia.org/wiki/Mel_scale
            const melMin = 2595 * Math.log10(1 + minFreq / 700);
            const melMax = 2595 * Math.log10(1 + maxFreq / 700);
            const melStep = (melMax - melMin) / barCount;
            const melValue = melMin + i * melStep;
            const startFreq = 700 * (Math.pow(10, melValue / 2595) - 1);
            const endFreq = 700 * (Math.pow(10, (melValue + melStep) / 2595) - 1);
            return {
                startFreq: startFreq,
                endFreq: endFreq,
                smoothValue: 0
            };
        });
        const getFrequencyBinIndex = (frequency)=>{
            const nyquist = audioContext.sampleRate / 2;
            return Math.round(frequency / nyquist * (analyser.frequencyBinCount - 1));
        };
        function drawSpectrum() {
            analyser.getByteFrequencyData(frequencyData);
            canvasCtx.clearRect(0, 0, canvas.width / scaleFactor, canvas.height / scaleFactor);
            canvasCtx.fillStyle = backgroundColor;
            canvasCtx.fillRect(0, 0, canvas.width / scaleFactor, canvas.height / scaleFactor);
            let isActive = false;
            const totalBarsWidth = bands.length * barWidth + (bands.length - 1) * barGap;
            const startX = (canvas.width / scaleFactor - totalBarsWidth) / 2; // Center bars
            const adjustedCircleRadius = barWidth / 2; // Fixed radius for reset circles
            bands.forEach((band, i)=>{
                const startIndex = getFrequencyBinIndex(band.startFreq);
                const endIndex = getFrequencyBinIndex(band.endFreq);
                const bandData = frequencyData.slice(startIndex, endIndex);
                const bandValue = bandData.reduce((acc, val)=>acc + val, 0) / bandData.length;
                const smoothingFactor = 0.2;
                if (bandValue < 1) band.smoothValue = Math.max(band.smoothValue - smoothingFactor * 5, 0);
                else {
                    band.smoothValue = band.smoothValue + (bandValue - band.smoothValue) * smoothingFactor;
                    isActive = true;
                }
                const x = startX + i * (barWidth + barGap);
                // Calculate bar height with a maximum cap
                const minHeight = 0;
                const barHeight = Math.max(minHeight, Math.min(band.smoothValue / 255 * barMaxHeight, barMaxHeight));
                let yTop, yBottom;
                const canvasHeight = canvas.height / scaleFactor;
                switch(barOrigin){
                    case "top":
                        yTop = adjustedCircleRadius;
                        yBottom = Math.min(adjustedCircleRadius + barHeight, canvasHeight - adjustedCircleRadius);
                        break;
                    case "bottom":
                        yBottom = canvasHeight - adjustedCircleRadius;
                        yTop = Math.max(yBottom - barHeight, adjustedCircleRadius);
                        break;
                    case "center":
                    default:
                        yTop = Math.max(canvasHeight / 2 - barHeight / 2, adjustedCircleRadius);
                        yBottom = Math.min(canvasHeight / 2 + barHeight / 2, canvasHeight - adjustedCircleRadius);
                        break;
                }
                if (band.smoothValue > 0) {
                    canvasCtx.beginPath();
                    canvasCtx.moveTo(x + barWidth / 2, yTop);
                    canvasCtx.lineTo(x + barWidth / 2, yBottom);
                    canvasCtx.lineWidth = barWidth;
                    canvasCtx.strokeStyle = barColor;
                    canvasCtx.stroke();
                } else drawInactiveCircle(adjustedCircleRadius, barColor, x, yTop);
            });
            if (!isActive) drawInactiveCircles(adjustedCircleRadius, barColor);
            requestAnimationFrame(drawSpectrum);
        }
        function drawInactiveCircle(circleRadius, color, x, y) {
            switch(barLineCap){
                case "square":
                    canvasCtx.fillStyle = color;
                    canvasCtx.fillRect(x + barWidth / 2 - circleRadius, y - circleRadius, circleRadius * 2, circleRadius * 2);
                    break;
                case "round":
                default:
                    canvasCtx.beginPath();
                    canvasCtx.arc(x + barWidth / 2, y, circleRadius, 0, 2 * Math.PI);
                    canvasCtx.fillStyle = color;
                    canvasCtx.fill();
                    canvasCtx.closePath();
                    break;
            }
        }
        function drawInactiveCircles(circleRadius, color) {
            const totalBarsWidth = bands.length * barWidth + (bands.length - 1) * barGap;
            const startX = (canvas.width / scaleFactor - totalBarsWidth) / 2;
            const canvasHeight = canvas.height / scaleFactor;
            let y;
            switch(barOrigin){
                case "top":
                    y = circleRadius;
                    break;
                case "bottom":
                    y = canvasHeight - circleRadius;
                    break;
                case "center":
                default:
                    y = canvasHeight / 2;
                    break;
            }
            bands.forEach((_, i)=>{
                const x = startX + i * (barWidth + barGap);
                drawInactiveCircle(circleRadius, color, x, y);
            });
        }
        drawSpectrum();
        // Handle resizing
        window.addEventListener("resize", resizeCanvas);
        return ()=>{
            audioContext.close();
            window.removeEventListener("resize", resizeCanvas);
        };
    }, [
        backgroundColor,
        barColor,
        barCount,
        barGap,
        barLineCap,
        barMaxHeight,
        barOrigin,
        barWidth,
        track
    ]);
    return (0, $5Zyvw$reactjsxruntime.jsx)("canvas", {
        ref: canvasRef,
        style: {
            display: "block",
            width: "100%",
            height: "100%"
        }
    });
});
$a1dfa75b13e6bb9b$export$59bf27bd43679db6.displayName = "VoiceVisualizer";
